kratos/processes/tetrahedral_mesh_orientation_check.cpp

```diff
@@ -0,0 +1,267 @@
+                // localindex_node_on_face the local_index of one of the nodes on the face
+                unsigned int counter = 0;
+                for(unsigned int i=0; i<rGeom.size(); i++)
+                {
+                    if(i != outer_node_index)
+                    {
+                        aux[counter++] = rGeom[i].Id();
+                        localindex_node_on_face = i;
+                    }
+                }
+
+                //*** THE ARRAY OF IDS MUST BE ORDERED!!! ***
+                std::sort(aux.begin(), aux.end());
+
+                hashmap::iterator it_face = faces_map.find(aux);
+                if(it_face != faces_map.end() ) // It was actually found!!
+                {
+                    // Mark the condition as visited. This will be useful for a check at the endif
+                    std::vector<Condition::Pointer>& list_conditions = it_face->second;
+                    for (Condition::Pointer p_cond : list_conditions) {
+                        p_cond->Set(VISITED,true);
+                    }
+
+                    if(mrOptions.Is(ASSIGN_NEIGHBOUR_ELEMENTS_TO_CONDITIONS))
+                    {
+                        WeakPointerVector< Element > VectorOfNeighbours;
+                        VectorOfNeighbours.resize(1);
+                        VectorOfNeighbours(0) = Element::WeakPointer( *it_elem.base() );
+                        for (Condition::Pointer p_cond : list_conditions) {
+                            p_cond->SetValue(NEIGHBOUR_ELEMENTS, VectorOfNeighbours);
+                        }
+                    }
+
+                    // Compute the normal of the face
+                    array_1d<double,3> face_normal = ZeroVector(3);
+                    GeometryType& r_face_geom = (list_conditions[0])->GetGeometry(); // The geometry is shared, we just take the first one
+
+                    Point::CoordinatesArrayType local_coords;
+                    local_coords.clear();
+                    noalias(face_normal) = r_face_geom.Normal(local_coords);
+
+                    // Do a dotproduct with the DenseVector that goes from
+                    // "outer_node_index" to any of the nodes in aux;
+                    array_1d<double,3> lvec = rGeom[outer_node_index]-rGeom[localindex_node_on_face];
+
+                    const double dotprod = inner_prod(lvec, face_normal);
+
+                    // If dotprod > 0 then the normal to the face goes in the same half space as
+                    // an edge that goes from the space to the node not on the face hence the face need to be swapped
+                    if(dotprod > 0) {
+                        r_face_geom(0).swap(r_face_geom(1));
+                        face_normal = -face_normal;
+
+                        CondSwitchCount++;
+                    }
+
+                    if(mrOptions.Is(COMPUTE_NODAL_NORMALS)) {
+                        double factor = 1.0/static_cast<double>(r_face_geom.size());
```

Comment:
this is not correct. the normal will not be proportional to the area, and this will not work for general (non triangular) geometries

Reply:
True, I just compied the code of the current header https://github.com/KratosMultiphysics/Kratos/blob/051578a0477f41c95e7f4317d5b1f4dcdb977bd0/kratos/processes/tetrahedral_mesh_orientation_check.h#L115 to a cpp, I can update it